crates/in_stream/src/mem.rs

```diff
@@ -0,0 +1,433 @@
+    }
+
+    /// private constructor, you probably want `bind`
+    fn priv_new(url: Url2, recv: crossbeam_channel::Receiver<InStreamMem>) -> Self {
+        Self {
+            url,
+            recv,
+            accept_queue: Vec::new(),
+        }
+    }
+}
+
+impl Drop for InStreamListenerMem {
+    fn drop(&mut self) {
+        get_mem_manager().unbind(&self.url);
+    }
+}
+
+impl InStreamListener<&mut [u8], &[u8]> for InStreamListenerMem {
+    type Stream = InStreamMem;
+
+    fn raw_bind<C: InStreamConfig>(url: &Url2, config: C) -> Result<Self> {
+        let _ = MemBindConfig::from_gen(config)?;
+        get_mem_manager().bind(url)
+    }
+
+    fn binding(&self) -> Url2 {
+        self.url.clone()
+    }
+
+    fn accept(&mut self) -> Result<<Self as InStreamListener<&mut [u8], &[u8]>>::Stream> {
+        loop {
+            // first, drain all pending connections from our recv channel
+            match self.recv.try_recv() {
+                Ok(stream) => {
+                    self.accept_queue.push(stream);
+                }
+                Err(crossbeam_channel::TryRecvError::Empty) => break,
+                Err(crossbeam_channel::TryRecvError::Disconnected) => {
+                    // wait until our user has accepted all pending connections
+                    // before letting them know the channel is broken
+                    if self.accept_queue.is_empty() {
+                        return Err(ErrorKind::BrokenPipe.into());
+                    }
+                }
+            }
+        }
+        if self.accept_queue.is_empty() {
+            // acceptor is non-blocking we have nothing to return
+            return Err(Error::with_would_block());
+        }
+        // pull the next item off the queue
+
+        Ok(self.accept_queue.remove(0))
+    }
+}
+
+impl InStreamListenerStd for InStreamListenerMem {
+    type StreamStd = InStreamMem;
+
+    fn accept_std(&mut self) -> Result<<Self as InStreamListenerStd>::StreamStd> {
+        self.accept()
+    }
+}
+
+#[derive(Debug)]
+/// memory stream specific connect config
+pub struct MemConnectConfig {}
+
+impl Default for MemConnectConfig {
+    fn default() -> Self {
+        Self {}
+    }
+}
+
+impl InStreamConfig for MemConnectConfig {}
+
+/// a singleton memory transport
+/// could be used for unit testing or for in-process ipc
+#[derive(Debug)]
+pub struct InStreamMem {
+    url: Url2,
+    send: Option<crossbeam_channel::Sender<Vec<u8>>>,
+    recv: Option<crossbeam_channel::Receiver<Vec<u8>>>,
+    recv_buf: Vec<u8>,
+}
+
+impl InStreamMem {
+    pub fn connect(url: &Url2, config: MemConnectConfig) -> Result<Self> {
+        InStreamMem::raw_connect(url, config)
+    }
+}
+
+impl InStream<&mut [u8], &[u8]> for InStreamMem {
+    /// we want a url like mem://
+    const URL_SCHEME: &'static str = SCHEME;
+
+    fn raw_connect<C: InStreamConfig>(url: &Url2, config: C) -> Result<Self> {
+        let _ = MemConnectConfig::from_gen(config)?;
```

Comment:
Confused by this pattern. Why is the return value ignored? How does it get associated with the memory manager? Is there a sneaky singleton underneath the hood?

Reply:
Don't actually need any configuration for memory connects, so it's literally ignored. Just wanted the placeholder / error case to be in there in case we add configuration elements in the future.